### DISCUSSION TITLE
Users are receiving expired OTPs

### DISCUSSION THREAD

**Discussion by Neuramare** [supabase/supabase#29686] [Questions]
Some of our users (50%) cannot login because they are receiving expired OTPs 🥲 

Some portions of my code (React Native):
export const supabase = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    storage: AsyncStorage,
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: false,
  },
});

const { data, error } = await supabase.auth.signInWithOtp({
      email: email,
    });

const { error } = await supabase.auth.verifyOtp({
      email: email,
      token: otp,
      type: "email",
    });

Package: 
"@supabase/supabase-js": "^2.43.4"

Event message from /verify EP:
{"component":"api","duration":3454168,"error_code":"otp_expired","level":"info","method":"POST","msg":"request completed","path":"/verify","referer":"http://localhost:3000","remote_addr":"8********51","request_id":"8cd4b69d17c0b14f-MAN","status":403,"time":"2024-10-04T11:08:19Z"}

We started having otp_expired issues at:
2024-10-02T19:10:12.000Z

On my project it's happening for 50% of the users. We usually verify 80 tokens / hour.

**Comment by savelichalex** [comment]
Experience the same on my project, we have very low number of requests less than 50 per day.

**Reply by BidMyTree** [reply]
Yes, same here. Very annoying and I have no idea how to fix it. My site kicks me out in less than hour.

**Comment by Neuramare** [comment]
Hi @savelichalex , I hope someone will take care of this very soon.

We are already 3 customers having this issue, feel free to check also on Discord in case someone from Supabase replies:
https://discord.com/channels/839993398554656828/1291708783009202176

**Comment by Neuramare** [comment]
![Screen Shot 2024-10-04 at 13 20 39 PM](https://github.com/user-attachments/assets/c68b5264-c69d-4cb5-b2e1-bbd9120bb7f3)
![Screen Shot 2024-10-04 at 13 58 45 PM](https://github.com/user-attachments/assets/be9b3394-2839-4d2d-be4f-65562230b6db)

Still sending expired OTPs

**Comment by Neuramare** [comment]
Hi @GaryAustin1 , could you please help us?

**Reply by GaryAustin1** [reply]
Hopefully you have support requests in also.  Seeing several reports on Discord too.  Support would be the one to escalate to the dev team if it is a broad issue.

**Comment by Neuramare** [comment]
![Screen Shot 2024-10-04 at 19 36 43 PM](https://github.com/user-attachments/assets/3b18f2c4-6635-457a-85c9-f3ac13ec7592)

I really can't see any logic behind it. Some requests made to /verify do work and some don't.

**Comment by Neuramare** [comment]
I've created a PR here, since it's just a test maybe it raises the issue:
https://github.com/supabase/auth/pull/1789

I did not setup the project on my machine, someone can run the test and check if it passes?

**Comment by hf** [comment]
We haven't really touched this code in a while so I'm not able to identify the root cause. 

Could it be that your users are facing the email prefetch link issue, where certain email clients visit the links and thus render them useless?

**Reply by GaryAustin1** [reply]
Working with the 3 users  in Discord also.  2 of them insist there is no link in the emails.
https://discord.com/channels/839993398554656828/1291708783009202176  long thread with testing.

Is it possible auth has not been upgraded in awhile for them until the force 2 days ago.   I could not answer how often you force auth upgrades versus doing an infrastructure update pulls it.  Issues all seemed to start 2 days ago.

**Reply by Neuramare** [reply]
Unfortunately the problem is not the prefetch link, since we are sending OTPs to users and not magic links.
As you can see I have this problem since 2 October, the other dev on the Discord channel since 3 October.
Around 20/40 users cannot login every hour. I have 17k users and 1600 MAU.

A lot of users are affected, it's not an edge case.

![Screen Shot 2024-10-05 at 18 25 11 PM](https://github.com/user-attachments/assets/816c64c3-86df-4f2a-b591-7d9f431574e6)

**Reply by DanielHeu** [reply]
please see my comment below

**Comment by DanielHeu** [comment]
I just discovered that it seems to be a case sensitive kind of problem.
My account-email in the database is daniel@healactively.com (lower case d), if I now request an otp with Daniel@healactively.com, I get an otp send to my email. But If I now try to verify this otp with Daniel@healactively.com (upper case D), I get otp is invalid. 
If request the otp with daniel@healactively.com and later verify it with daniel.healactively.com, everything works fine.

This started to happen at the 3rd of october

**Reply by Neuramare** [reply]
Same happens to me, but I have some users with normal emails (without uppercases) receiving invalid OTPs.

**Reply by DanielHeu** [reply]
it looks like all emails are stored lowercase in the users table, so this could also happen on signup. If the user signs-up with a non lower case email like Daniel@, the otp will be veryfied against Daniel@ but in the database the user is stored with daniel@, so the verify method will get back not valid

**Reply by DanielHeu** [reply]
and in the auth.one_time_token table the email is also in all lower case

**Comment by Neuramare** [comment]
https://github.com/supabase/auth/blob/410b8acdd659fc4c929fe57a9e9dba4c76da305d/internal/crypto/crypto.go#L50
func GenerateTokenHash(emailOrPhone, otp string) string {
	return fmt.Sprintf("%x", sha256.Sum224([]byte(emailOrPhone+otp)))
}

I tried running this function on 2 different email addresses and their OTP:

Case 1 (not matching):
email: Angelo.lamonaca@gmail.com
otp: 747447
Generated token hash: b6e39fc5938559c3a6f3a383330269ebfe485c11799e872bc033d19b
Token Hash in the one_time_token table: pkce_a72505002d8aa2e411d8ba3a5ed075f8a4b8626dbd8d1b0b2d4ed7aa

Case 2 (matching):
email: admin@neuramare.com
otp: 432192
Generated token hash: 4b9333fd7e069cdc2b1d143d45866a9a273a78ea3320b7ae74b6a450
Token Hash in the one_time_token table: pkce_4b9333fd7e069cdc2b1d143d45866a9a273a78ea3320b7ae74b6a450

**Reply by Neuramare** [reply]
I see the problem, in Case 1 the token was generated without considering the uppercase, in fact if you generate the token hash with angelo.lamonaca@gmail.com and 747447, you get the hash saved in the table:
a72505002d8aa2e411d8ba3a5ed075f8a4b8626dbd8d1b0b2d4ed7aa

**Reply by DanielHeu** [reply]
So the token hash is always created with the lower case email. It looks like the verify function somehow stopped converting the emails to lowercase before generating the hash and checking it.
Worked before the 3rd of october, so it must be a recent change that caused this

**Reply by Neuramare** [reply]
Yes, but it's the first time I check the auth service so I don't know if this is the real issue, I'm just assuming. Hopefully @hf will figure out the problem.

**Reply by hf** [reply]
Yes this was the issue, we've deployed the fix.

**Reply by baxuu** [reply]
@hf i still have this problem. when the email address is capitalized, I get an error about an invalid token

**Comment by sfarkas1988** [comment]
Is this issue solved?
I experience the same right now and I suppose the issue is because my client is in the US, we are in Germany, trying to login with otp.
If the hash has a relation to a time and the timezone might not be handles properly, the hash can differ! 
Is that maybe an issue here ?

**Reply by GaryAustin1** [reply]
Yes you probably have some other issue.
Probably should start a new post or try Discord.

**Comment by nextglabs** [comment]
Experiencing a similar issue with certain users ...

**Comment by StacySiz** [comment]
I'm having the same issue — the OTP code expires immediately after I click the confirmation email, even though the expiration time is set to the standard duration.

Any fix?

**Comment by guillaumehcht** [comment]
Same issue here, what's going on with this? Any solutions?

**Comment by ONLYstcm** [comment]
I set up auth with emails today and I'm getting the same error. Every OTP code I try is expired for some reason

**Comment by BidMyTree** [comment]
Yes, big problem all the time.

**Comment by frederickobrien** [comment]
I'm getting this issue as well, though oddly for some email addresses it works fine and others consistently fails.

**Comment by pourya7** [comment]
I faced this issue yesterday and the following changes fixed it for us:

before:
```
// type is set on the link coming from email template
const type = searchParams.get('type') as EmailOtpType | null
...
const { error } = await supabase.auth.verifyOtp({
  type,
  token_hash,
})
```

after:
```
 const { error } = await supabase.auth.verifyOtp({
  type: 'email',
  token_hash,
})
```

This is briefly mentioned in the documentation: https://supabase.com/docs/reference/javascript/auth-verifyotp
The `magicLink` and `signup` types are deprecated.

**Comment by kyoshino** [comment]
I’m also experiencing the issue. Solved by switching to a different VPN server location. Maybe Supabase is blocking some IPs?

**Comment by zhihengGet** [comment]
it is probably gmail prefetching the link
